lib/WUI/automata/core.cpp

```diff
@@ -0,0 +1,143 @@
+#include "core.h"
+
+#include <cassert>
+#include <cctype>
+
+using std::nullopt;
+using std::optional;
+
+namespace {
+
+typedef bool (*SpecialCheck)(uint8_t);
+
+bool anything(uint8_t) {
+    return true;
+}
+
+bool whitespace(uint8_t byte) {
+    return isspace(byte);
+}
+
+bool horiz_space(uint8_t byte) {
+    return byte == ' ' or byte == '\t';
+}
+
+bool digit(uint8_t byte) {
+    return isdigit(byte);
+}
+
+SpecialCheck specials[] = {
+    anything,
+    whitespace,
+    horiz_space,
+    digit,
+};
+
+}
+
+namespace automata {
+
+bool Transition::matches(uint8_t byte) const {
+    switch (label_type) {
+    case LabelType::Char:
+        return byte == label;
+    case LabelType::CharNoCase:
+        return tolower(byte) == label;
+    case LabelType::Special:
+        return specials[label](byte);
+    default:
+        assert(0);
+        return false;
+    }
+}
+
+std::optional<Event> Automaton::gen_event(StateIdx old_state, StateIdx new_state, uint8_t payload) const {
+    const State &o = states[old_state];
+    const bool leave = o.emit_leave;
+    const State &n = states[new_state];
+    const bool enter = n.emit_enter;
+
+    if (leave || enter) {
+        return Event {
+            old_state,
+            new_state,
+            leave,
+            enter,
+            payload,
+        };
+    } else {
+        return nullopt;
+    }
+}
+
+std::optional<TransitionResult> Automaton::transition(StateId old, uint8_t byte) const {
+    const State &state = states[old.state];
+    if (state.has_path) {
+        const uint8_t converted = state.path_nocase ? tolower(byte) : byte;
+        const char *path = paths[state.path].value;
+        if (path[old.path] == converted) {
+            if (path[old.path + 1] == '\0') {
+                StateId new_state(old.state + 1);
+                return TransitionResult {
+                    new_state,
+                    gen_event(old.state, new_state.state, byte),
+                    false,
+                };
+            } else {
+                return TransitionResult { old.path_step(), nullopt, false };
+            }
+        }
+    }
+
+    const State &sentinel = states[old.state + 1];
```

Comment:
I know it looks a bit weird and it sounds strange, but I actually need this behavior. Let me explain why.

The reason why paths even exist is to save on number of states when there's a constant string (I'd like to use them for tries eventually too, and save even more). Otherwise, the number of states would be bigger and the flash consumption higher. This is to recognize things like known headers or known methods.

Now, when there's an known header, it can _start_ as something we know, but get out of that path in the middle. In that case, I can't just give up the whole parsing. I have to go to some kind of „Unknown header“ state. So I have to check for that state.

The way these things are used is that the path state has some wildcard/fallback transitions.

Maybe we should have some kind of states from the state and states from the path. I don't know if I want to introduce them right now, though.

Reply:
Ok, get it. Those are a bit rough edges rn, but something we can focus on later :)

And when the time comes, I would also be down to use another name for it (not "path"). Paths already exist in graph theory, and it took me a while to get my head around the fact that they mean here something slightly different :). But I didn't come up with a good name yet (it is something like "multi transition" here, not a great name either 😁)

Reply:
Btw, one great (and super simple to implement) feature would be outputting the graphs in DOT lang for easy visualisation with Graphviz 🤓